### ISSUE TITLE
Deleting and Recovering of Bookmarks and Tracks

### ISSUE THREAD

**Issue by kirylkaveryn** [organicmaps/organicmaps#9859]
## 1. Problem Statement
Users need the ability to delete lists, bookmarks, and tracks within the app using various UI actions and buttons across different screens. The deletion process should be consistent, simple, and support a recovery mechanism to restore accidentally deleted elements.

#### Current Implementation:
- Lists Deletion: Lists are moved to a ./Trash directory that is not cloud-synced.
- Bookmark Recovery: A fast, in-place bookmark recovery option allows restoring an accidentally deleted file directly from the “Place Page” screen (available until the screen is closed).

## 2. Key Requirements
- Deletion: The process to delete a list, bookmark, or track should be simple, without requiring confirmation dialogs from every app place.
- Recovery: Deleted elements must be recoverable from the “Recently Deleted” screen and restored to their original source lists.

#### Steps:
1. Identify the element (track or bookmark) to delete.
2. Add a <Deleted> tag or an attribute (e.g., <Placemark deleted="true">).

#### Pros:
- Simplifies the recovery process since elements remain in the same file.
- Eliminates the need for additional file management.
- Maintains data structure integrity.
- Allows cloud-synced deleted content to remain accessible across all devices and platforms.

#### Cons:
- Increases the original file size due to retained data.
- Adds parsing overhead for systems processing the KML.
- Risks cluttering the file with unnecessary metadata.
- Deleted elements may not be parsed correctly by other applications or services.

#### Steps:
Extract the element (track or bookmark) from the original KML file.
Create or append the extracted element to a secondary KML file in the ./Trash directory.

#### Pros:
- Keeps the original file clean and reduces its size.
- Separates active and deleted data for better clarity.
- Deleted elements are only processed when the user views the “Recently Deleted” screen.
#### Cons:
- Recovery involves reading two files, increasing complexity.
- Secondary files in ./Trash could be lost if not managed properly.
- Requires additional file-handling logic.
- Harder to link source and deleted versions in cases of manual file renaming or deletion.
- Deleted elements in ./Trash are not synced across devices.
- If both source and deleted_ versions of a file exist in ./Trash, additional logic must resolve conflicts to maintain consistency.

#### Steps:
1. Extract the element (track or bookmark) from the original KML file.
2. Append the deleted element to a centralized audit KML file with metadata linking it to its original group.
3. Replace the deleted element in the original file with a lightweight reference placeholder (e.g., <Placemark ref="auditFile.kml#ID123" />).

#### Pros:
- Keeps the original file clean and reduces its size.
- Reduces file count by consolidating all deleted data into one audit KML file.
- Lightweight references minimize the size of the original file while maintaining traceability.

#### Cons:
- Requires additional serialization/deserialization logic for the audit KML file.
- Recovery involves resolving references from the original file to the audit file.
- Increases implementation complexity.

**Comment by pastk** [comment]
> ### Solution 1: Tagging Deleted Elements in Original KML
> 
> #### Cons:
> 
>     * Increases the original file size due to retained data.
> 
>     * Adds parsing overhead for systems processing the KML.
> 
>     * Risks cluttering the file with unnecessary metadata.
> 

I think these are minor concerns as its unlikely that a ratio of deleted elements would be high.
Given there should be some kind of "trash collection". E.g. keep for 30 days by default.
The question is how/when the cleanup should be triggered.
Probably each file should have a "last check" timestamp...

>     * Deleted elements may not be parsed correctly by other applications or services.
> 

When a user exports bookmarks the export process should produce a clean file without deleted elements.

> 
> ### Solution 2: Moving Elements to a Secondary KML File linked by the file name with the source file
>
> #### Cons:
> 
>     * Harder to link source and deleted versions in cases of manual file renaming or deletion.

I think its fine to omit this case for simplicity. If a source file was renamed then its trash is lost.

> 
>     * Deleted elements in ./Trash are not synced across devices.

I think its logical and expected?

> 
>     * If both source and deleted_ versions of a file exist in ./Trash, additional logic must resolve conflicts to maintain consistency.

If the whole source file was deleted then just remove its trash immediately.

How to implement trash collection here?

> 
> ### Solution 3: Moving Elements to the Audit KML File that tracks all the deleted content
>  

This option might be the easiest for trash collection (only one file to keep a track of).

**Comment by pastk** [comment]
From user's POV / UX its all the same.

So makes sense to pick a solution which is easiest to implement and maintain..

**Comment by pastk** [comment]
> * Deletion: The process to delete a list, bookmark, or track should be simple, without requiring confirmation dialogs from every app place.
> 
> * Recovery: Deleted elements must be recoverable from the “Recently Deleted” screen and restored to their original source lists.

Do any other map apps follow a "trash bin" concept?

OM strives for simplicity. Maybe we're overcomplicating here and its better to cope with a confirmation dialog for valuable / hard to restore data (tracks and whole lists)? The confirmation dialog is a classic solution and for not-very-often deletion cases (lists and tracks) its added UX overhead might be preferrable to a "trash bin" solution, which is complex both UX-wise and implementation-wise.

**Comment by biodranik** [comment]
If tracks will be always automatically saved without any intermediate 3-choice menu, with a follow-up summary dialog about a track and a delete button on it, then deletion will be a frequent operation and confirmation may be annoying.

Bin may be okay for whole lists with a lot of valuable info. Restoring accidentally deleted bookmark or track can be done independently from a bin.

**Comment by pastk** [comment]
> If tracks will be always automatically saved without any intermediate 3-choice menu, with a follow-up summary dialog about a track and a delete button on it, then deletion will be a frequent operation and confirmation may be annoying.

Getting rid of the save dialog will not affect frequency of deletions, of course.

The only difference is that a save dialog is modal and requires a user to make a choice (save or delete) while auto-save makes the delete more optional (and hence a little more safe from accidental use).

**Comment by biodranik** [comment]
Right, I wasn't clear. I meant the delete button on the track's PP will be used way more frequently.

**Comment by kirylkaveryn** [comment]
> > ### Solution 3: Moving Elements to the Audit KML File that tracks all the deleted content
> 
> This option might be the easiest for trash collection (only one file to keep a track of).

This on seems like the hardest to implement, because there should be a lot of updates in kml serdes operations. To track all the references from the deleted element to the groups.

> Do any other map apps follow a "trash bin" concept?

Usually, the maps apps use confirmation dialogs to handle such situations. But the trash bin concept is widely used in a lot of apps with user-generated content that may be deleted frequently.
